dbms/src/main/java/org/polypheny/db/routing/routers/AbstractDqlRouter.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2019-2021 The Polypheny Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.polypheny.db.routing.routers;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import java.util.Collections;
+import java.util.List;
+import lombok.extern.slf4j.Slf4j;
+import org.polypheny.db.catalog.entity.CatalogTable;
+import org.polypheny.db.plan.RelOptCluster;
+import org.polypheny.db.prepare.RelOptTableImpl;
+import org.polypheny.db.rel.RelNode;
+import org.polypheny.db.rel.RelRoot;
+import org.polypheny.db.rel.core.ConditionalExecute;
+import org.polypheny.db.rel.core.SetOp;
+import org.polypheny.db.rel.logical.LogicalTableModify;
+import org.polypheny.db.rel.logical.LogicalTableScan;
+import org.polypheny.db.rel.logical.LogicalValues;
+import org.polypheny.db.routing.LogicalQueryInformation;
+import org.polypheny.db.routing.Router;
+import org.polypheny.db.schema.LogicalTable;
+import org.polypheny.db.tools.RoutedRelBuilder;
+import org.polypheny.db.transaction.Statement;
+
+
+/**
+ * Abstract router for DQL routing.
+ * It will create rel builders and select column placements per table.
+ * There are three abstract methods:
+ * handleHorizontalPartitioning
+ * handleVerticalPartitioning
+ * handleNonePartitioning
+ * Every table will be check by the following order:
+ * first if partitioned horizontally -> handleHorizontalPartitioning called.
+ * second if partitioned vertically or replicated -> handleVerticalPartitioningOrReplication
+ * third, all other cases -> handleNonePartitioning
+ */
+@Slf4j
+public abstract class AbstractDqlRouter extends BaseRouter implements Router {
+
+    /**
+     * Boolean which cancel routing plan generation.
+     * This is useful if implementing router can not propose plans in every situation.
+     */
+    protected boolean cancelQuery = false;
```

Comment:
How does this work? Why is it required?

Reply:
First of all, the main logic of how a router creates a plan has not changed (yet), and during traversing the relnode 
router is called for every tablePlacement. 

In Universal Routing, not every router needs to propose a plan for every query. But, every router will be asked to try to prepare a plan for the query. If a router sees, that he is not able to route (maybe after tablePlacement 2 out of 4, the router is expected to return an empty list of proposed routing plans. To "abort" routing plan creation, the boolean cancelQuery is introduced and will be checked during creation. As we are in the middle of traversing a RelNode, this is the new to "abort" travsering the RelNode.

Maybe you have better ideas but it is just a mechanism, to not traverse the RelNode til the end.